### PR TITLE
treewide: fix pg_config / postgresql headers moved to dev output

### PR DESCRIPTION
**Pull request by wolfgangwalther**
This was supposed to happen in #294504, but the commit was accidentally left out when splitting off some libpq-related changes. Originated in #179962, by Sandro.

@Ma27 

ToDo:
- [x] Fix some more direct references to `${postgresql}`
- [x] Build all changed packages

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: pkgs/development/python-modules/psycopg2/default.nix [hunk 1]
```diff
@@ -37,7 +37,7 @@ buildPythonPackage rec {
     # some linker flags are added but the linker ignores them because they're incompatible
     # https://github.com/psycopg/psycopg2/blob/89005ac5b849c6428c05660b23c5a266c96e677d/setup.py
     substituteInPlace setup.py \
-      --replace "self.pg_config_exe = self.build_ext.pg_config" 'self.pg_config_exe = "${lib.getExe' buildPackages.postgresql "pg_config"}"'
+      --replace "self.pg_config_exe = self.build_ext.pg_config" 'self.pg_config_exe = "${lib.getDev buildPackages.postgresql}/bin/pg_config"'
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/psycopg2/default.nix:40]: We should replace this with --replace-fail to be on the safe side

File: pkgs/servers/sql/postgresql/ext/tsja.nix [hunk 2]
```diff
@@ -18,7 +18,7 @@ stdenv.mkDerivation rec {
 
   postPatch = ''
     substituteInPlace Makefile \
-      --replace /usr/local/pgsql ${postgresql} \
+      --replace /usr/local/pgsql ${lib.getDev postgresql} \
```
Comment (SuperSandro2000) [inline pkgs/servers/sql/postgresql/ext/tsja.nix:21]: Same here

**Comment by Ma27** [comment]
Damn it, sorry for missing that on the final iterations! :(

File: pkgs/servers/sql/postgresql/ext/tsja.nix (hunk 2 above)
Comment (Ma27) [reply pkgs/servers/sql/postgresql/ext/tsja.nix:21]: fwiw I won't block fixing staging regressions on changing untouched code parts.

**Comment by Ma27** [comment]
Built the following things so far, with success:
* PHP 8.4
* `python3.pkgs.psycopg2`
* `postgresqlPackages.tsja`


Can take a look into

```
ERROR:  Error installing /nix/store/8ms1n7ms7acp8xqcsg7bdqv8bsjbjd8k-pg-1.2.3.gem:
        ERROR: Failed to build gem native extension.

    current directory: /nix/store/zv1lgkq0aza3zap1xzry95cd53973nam-ruby3.3-pg-1.2.3/lib/ruby/gems/3.3.0/gems/pg-1.2.3/ext
/nix/store/r0ypxm1x0jf2v1837j7kcz56kjsc7rkp-ruby-3.3.4/bin/ruby extconf.rb --with-pg-config\=/nix/store/674zd30h0laaraw77a1c4pw0av55hd08-postgresql-16.4-dev/bin/pg_config
Using config values from /nix/store/674zd30h0laaraw77a1c4pw0av55hd08-postgresql-16.4-dev/bin/pg_config
```

from `/nix/store/v325dyz38dcqrpk58h20sq43xabyykpr-ruby3.3-pg-1.2.3.drv` (while building `zammod`) now.

File: pkgs/development/python-modules/psycopg/default.nix [hunk 3]
```diff
@@ -74,11 +74,17 @@ let
 
     nativeBuildInputs = [
       cython
+      # needed to find pg_config with strictDeps
+      # TODO: switch to pkg-config upstream
```
Comment (mweinelt) [inline pkgs/development/python-modules/psycopg/default.nix:78]: For whom is that TODO? I'm not going to do it, and I don't know whether upstream intends to.

**Review by Ma27 — APPROVED** [review]

**Pull request merged**